### DISCUSSION TITLE
Dexie Cloud Pricing

### DISCUSSION THREAD

**Discussion by b-straub** [dexie/Dexie.js#1735] [Q&A]
I don't grasp entirely the target audience for a Dexie Cloud driven application/site.
For a webapp the user based pricing doesn't seem to be very suitable, since for such apps usually only a subset of the installed user base is using the database simultaneously. Usually such cloud database pricing models are space/transaction based and post charged (e.g. Firebase, Azure).

Since I develop the .NET wrapper around Dexie, my perspective is driven by Webassembly apps, where an easy to uses cloud database especially with offline/online sync makes a lot of sense.

E.g. would could be the pricing for a shared shopping list app, does such an app even make sense?
Perhaps I miss something and the use case are more B2B apps with a defined number of users.

**Comment by dfahlander** [comment]
Thanks for bringing it up and sorry for a late reply! A shared shopping list would be a perfect example of an app that could use dexie cloud. I'm primarily trying to target app developers that build apps for end users. The current pricing model aims at making it cheap to get started building an app and grow its users, and also let it be easier for the customer to know the actual cost instead of guessing performance measurements such as CPU usage, network consumption, memory usage, storage etc.

I'm open to all ideas around pricing. The purpose is having a model, an API etc that makes it easy for app developers to launch and succeed with their app. We've been discussing to have a built-in trial support to help app developers get many trial users on board easily without a cost. If building a freemium app where only a few percent of the user will upgrade to a paid version of the app, it would be costly having to pay for all the users that aren't paying. For that reason, one of the features that could be included in a paid version would naturally be being able to sync and share data. A shared shopping list would be a perfect example - when the trial ends - they'll be able to keep using their app, edit and add items, but it would simply stop syncing until they upgrade. And the upgrade has to be done between end user and customer - so that customer can call some API in dexie cloud to resume the user's sync when user has paid to customer - and the user would start being counted in the purchase from Dexie Cloud to customer.

There's also always the [on-prem license option](https://dexie.org/cloud/docs/premium-software) with no limits of number of users.

**Reply by b-straub** [reply]
I think there are still missing parts or I didn't catch something. Let's create some real world example to get more clarity.

1) Dev wants to build a shopping list app, his customers will use the app on (multiple) devices and regularly share some shopping lists with other users. Dev don't want to be bound to app stores and mobile devices, therefore he chose to use a PWA and DexieCloud as backend. Dev has no payment or distribution service in place yet. The app should have a freemium pricing model as well. The points "app store" and "not only mobile devices" are important, since otherwise Dev would go for a MAUI app with InApp billing [InAppBilling](https://github.com/jamesmontemagno/InAppBillingPlugin) and may be Firebase or Azure as cloud sync.
When targeting iOS only, Dev could just use Apples infrastructure and CloudKit to sync and share with zero costs for an free or simple app.

- Dev will host the app as webassembly on his site, download is free
- Dev will  implement registration and InApp purchase, likely through a third party service since he don't want to deal with all the payment complexity
- each customer can test the app 14 days, including sync and sharing with other customer of his app

The flow could be something like:

- Customer loads webassembly
- Costumer needs to register, Dev issues a 14 day trial license allowing to login with DexieCloud
- Customer find share targets aka other Customers of Devs app with DexieCloud login, see #1668 
- after 14 days trial license expires and Dev needs to notify DexieCloud that login is not allowed anymore if customer has not bought a permanent license.

With the current implementation state Dev can neither prevent that the customer to going to login again with another UserID nor control that invitations for sharing will be send to people that are no customers of Devs app at all.

For Dev this will involve handing multiple payment models, registrations, etc. so very likely the target audience for DexieCloud will be quite small.

I think you should consider a radically different model that could create a unique selling point and allow OpenSource or small business developers to benefit from an open and easy to use cloud solution.

###  Open Source

- Dev publishes a PWA as Open Source
- users can use this app for free, if user want to sync across devices or share content with other users, he goes to DexieCloud and register his UserID and buy a subscription (likely a trial first). Given your current pricing model this can be made quite cheap.
- when users want to share content, DexieCloud will check if the invitations are send to some other users of the same app, if not, some error comes back and Dev can redirect to the users eMail with a nice message inviting the other user to install the app and register with DexieCloud.
- Money is all yours

### Small business

- Same as open source, DexieCloud subscriptions will be more expensive and you need to create a revenue model for the developer

### B2B

- Developer will buy subscription for n users from DexieCloud
- Developer register n userIDs with DexieCloud (not clear to me how to issue preregistration, is there a CLI call for, or do I need to upload member data)
- Developer sends out email how to download his B2B app and to login and with which userID

**Reply by b-straub** [reply]
Since Dexie Cloud is close to production now I have still no idea how to address a free PWA app.
An easy to use cross platform cloud sync will a be major benefit for free web applications. I hope you will include this in your pricing model, see my proposal from the comment before.

**Reply by dfahlander** [reply]
Thanks for reminder about this discussion. I'm open to discuss this but I'm not completely sure that I understand the requirements and whether you are sure that the current solution wouldn't work well with all the three scenarios as it is?

The current subscription solution is already deployed to all beta users. I have migrated all customers to it so all of you are using the Free version. But I have also extended every end user of every customer database to 90 active evaluation days. This is because no customer should get any disturbance in the service until the date and steps needed for production service are clear and communicated to all. This means that all end users can currently use the service in full and will be able to do so until their individual evaluation periods ends. New users get 30 days eval periods unless configured else in the [Dexie Cloud Manager (works but not officially released yet)](https://manager.dexie.cloud) .

As the subscription model works currently (see [blog post](https://medium.com/dexie-js/dexie-cloud-subscription-model-cbf9a709ce7), any open source project, Small Business or B2B could offer anyone to run their app for free (or paid - what they desire) and dexie cloud will allow any end user to use its full capabilities for an evaluation period of up to 30 days (the period is configurable in Dexie Cloud Manager) per end-user - not the Dexie Cloud subscription. If one of dev's end-users tries the app with sync and everything and the end-user evaluation eventually expires, end user will still be able to continue using the app in offline mode, without sync. Let's say they contact dev for a prod version to resume sync - then dev could manually or API-wise upgrade the end user from "eval" to "prod". This is done using the Dexie Cloud Manager or [the new /users endpoint in the REST API](https://dexie.org/cloud/docs/rest-api#users-endpoint). Free edition of Dexie Cloud only allows up to 3 end users to become "prod" instead of "eval". Being "prod" means there's no end date - subscription is ongoing until anything else are actively changed in manager or API. To upgrade more than 3 users, a subscription of Dexie Cloud Production needs to be purchased. When dev has purchased it, another 100 prod "seats" will be added, allowing dev to select up to 100 individual end users to be upgraded to "prod" accounts.

Dexie Cloud does not offer any payment solution for the customer directly with their dev's end users - it will be up to customer to sign-up with Stripe or any other provider to get paid according to their pricing and plan per user. It is up to the dev to upgrade individual end users to prod users. This is done using [Dexie Cloud Manager](https://manager.dexie.cloud) or [the new /users endpoint in the REST API](https://dexie.org/cloud/docs/rest-api#users-endpoint). A stripe webhook could typically be integrated to to upgrade the end user using the REST API.

We've had thoughts about offering subscriptions directly with end users and offer provision to our customers instead. Is this the proposal you are suggesting? Of course this could be very nice for the dev - not having to integrate with Stripe or anything - a bit like appstore / google play. But it would be another use case and it alone would not work for all either. It would also put more legal demands on dexie cloud. We're not there yet, but if this turns out to be a much requested feature we might do it in the future. And I would gladly continue to discuss it and get as much feedback as possible about how it should work.

**Reply by b-straub** [reply]
> We've had thoughts about offering subscriptions directly with end users and offer provision to our customers instead. Is this the proposal you are suggesting? Of course this could be very nice for the dev - not having to integrate with Stripe or anything - a bit like appstore / google play. But it would be another use case and it alone would not work for all either. It would also put more legal demands on dexie cloud. We're not there yet, but if this turns out to be a much requested feature we might do it in the future. And I would gladly continue to discuss it and get as much feedback as possible about how it should work.

Basically, that's the idea. The reason is simple. Personally, I'm mainly interested in free PWA applications. Some of my ideas require a database and often synchronisation. While for native apps the Android/iOS capabilities could provide such support for free, I don't know of any easy to integrate solution for web apps.

With end-user payment through Dexie Cloud, I could simply redirect the user to your payment process, and since for most users something like five production seats might be enough, you could offer something like an annual subscription at a reasonable price.  Such a solution may require some changes to the underlying authentication model, as the licence will be tied to a user's dedicated realm and define how many realm members they can synchronise with.

**Reply by dfahlander** [reply]
The idea I have been looking at is to provide a purchase solution for the end user (optinally available to redirect to for devs that don't want to provide their own solution), but let the dev set the price and my company gets a fair amount but the rest goes to the dev.

> With end-user payment through Dexie Cloud, I could simply redirect the user to your payment process, and since for most users something like five production seats might be enough, you could offer something like an annual subscription at a reasonable price. Such a solution may require some changes to the underlying authentication model, as the licence will be tied to a user's dedicated realm and define how many realm members they can synchronise with.

My guess would rather be that most users would buy single seats for themselves. Some small businesses might want to buy a few seats though - it all depends on the type of app their subscribing to. I strongly believe the app should be in focus and not sell some kind of generic storage - that might work if we were Google or Microsoft - but the particular end-user would only be interested in the app thei're looking at and buy a subscription for that.

**Reply by b-straub** [reply]
> The idea I have been looking at is to provide a purchase solution for the end user (optinally available to redirect to for devs that don't want to provide their own solution), but let the dev set the price and my company gets a fair amount but the rest goes to the dev.

Great idea, for FOSS the devs amount should go down to zero. The purchase will then be just for providing the cloud infrastructure and maintenance.

> My guess would rather be that most users would buy single seats for themselves.

One seat would allow synchronisation between multiple devices, right? Still having FOSS in mind here as well. Some solutions will require sharing a realm with a small group of people like family or friends.

**Reply by dfahlander** [reply]
> One seat would allow synchronisation between multiple devices, right?

Yes!

> Still having FOSS in mind here as well. Some solutions will require sharing a realm with a small group of people like family or friends.

I see. Makes sense.

**Reply by b-straub** [reply]
> I strongly believe the app should be in focus and not sell some kind of generic storage

Totally agree, but let's assume I want to make the ultimate shopping list app (FOSS or paid) since I'm not happy with the current solutions usability. Sure I want to share some of the shopping lists with my family or even sometimes with friends (e.g. for a mutual friend's birthday party).

So how this would work?

Buying three production seats for the family for daily usage and adding the friends as eval users for the birthday case. Eval doesn't seem the right choice here.

**Reply by dfahlander** [reply]
You'd probably share the link that leads to your list in the app to your friends - they sign-in and start their free evaluation period. If this is a short usage (single day or a few days) they might only consume a part of their evaluation and may continue like this for your next birthday until the eval days reaches zero. They would probably buy a subscription after becoming a recurring user for a while. Or they change to another app at that point. It would all depend on how great the app would be. The point is: already with the current model, it is easy to share apps and let anyone try it and use it for real until knowing if they're willing to continue using the app and pay for it.
